### ISSUE TITLE
Clearer docs on the service runner/container

### ISSUE THREAD

**Issue by mattbennett** [onefinestay/nameko#287]
c.f. https://github.com/onefinestay/nameko/issues/286

It's not helpful to show the ``stop()`` immediately after starting in https://github.com/onefinestay/nameko/blob/master/docs/examples/service_container.py#L16 and the runner example. Better to show ``wait()`` and point people towards the ``nameko.cli.run`` module for an importable function for running services programmatically.

**Comment by laike9m** [comment]
Hi matt, it seems neither invoking start() on container or runner will start service unless I call wait(). Is this expected?

**Comment by mattbennett** [comment]
@laike9m `start()` does start the service, but you must `wait()` on the container or the runner to stop your process or thread from exiting.

The best way to run a service is `nameko run` on the command line. Or, if you want to embed the service in another application, wrap the `run` function in `nameko.cli.run` -- https://github.com/onefinestay/nameko/blob/master/nameko/cli/run.py#L117 (which is what the command line tool uses under the hood)

**Comment by laike9m** [comment]
Thank you. I'm using a runner because in this way I'm able to stop the service programmatically. Actually there's another issue: I have to call `kill` to stop it while `stop` doesn't work.

**Comment by mattbennett** [comment]
Sorry, I forgot about this thread.

`stop()` waits for all running workers and managed threads to exit gracefully. You can't expose `container.stop()` in an entrypoint and expect to the container to stop when you call it -- this would create a deadlock between the container trying to stop and the worker or managed thread running the `stop()` method.

If you want to expose an entrypoint that is capable of stopping the container, you'd need to do it "out of band" (i.e. using a thread that wasn't created by the container)

Here is a toy example that does that:

``` python
from nameko.rpc import rpc
from nameko.extensions import DependencyProvider
import eventlet

class Stopper(DependencyProvider):

    def stop(self):
        # can't use container.spawn_managed_thread here either
        eventlet.spawn_n(self.container.stop)

    def get_dependency(self, worker_ctx):
        return self


class Service(object):
    name = "stop"

    stopper = Stopper()

    @rpc
    def stop(self):
        self.stopper.stop()
        return "stopping"
```

```
(tmp-6d6e28416a4d3fa9)Matts-13-inch-Macbook-Pro:tmp mattbennett$ nameko run stop
starting services: stop
Connected to amqp://guest:**@127.0.0.1:5672//
```

```
(tmp-6d6e28416a4d3fa9)Matts-13-inch-Macbook-Pro:~ mattbennett$ nameko shell
Nameko Python 2.7.10 (default, Jul  6 2015, 15:19:48)
[GCC 4.2.1 Compatible Apple LLVM 6.1.0 (clang-602.0.53)] shell on darwin
Broker: amqp://guest:guest@localhost
>>> n.rpc.stop.stop()
```

**Comment by davidszotten** [comment]
it seems unusual to expose an endpoint to remotely shut down a service. is this definitely what you want when you say "stop the service programatically". Typically this is done e.g. via a process manager like supervisor, which stops (or kills) the service via posix signals. E.g. `nameko run` listens for `SIGTERM` and calls `stop`

**Comment by mattbennett** [comment]
Yes indeed, the example above is not a demonstration of how you *should* achieve this. It is only meant to illustrate an "out of band" thread.

Generating a `SIGTERM` signal is a much better way to have the container to stop gracefully.

**Comment by WoLfulus** [comment]
Is there a method for dependecy stop/shutdown/cleanup?
